autokeras/hypermodel/preprocessor.py

```diff
@@ -142,67 +142,174 @@ def set_weights(self, weights):
+                self.sentence_containers[element] = 1
+            else:
+                self.sentence_containers[element] += 1
+        set4sentence.clear()
 
     def finalize(self):
-        self._texts = np.array(self._texts)
-        self.vectorizer.fit(self._texts)
-        data = self.vectorizer.transform(self._texts)
-        self.shape = data.shape[1:]
-        if self.targets:
-            self.selector = feature_selection.SelectKBest(
-                feature_selection.f_classif,
-                k=min(self.vectorizer.max_features, data.shape[1]))
-            self.selector.fit(data, self.targets)
+        self.stc_num += 1  # idf smoothing
+        for word in self.sentence_containers:
+            idf = (np.log((self.stc_num)/(self.sentence_containers[word] + 1)) + 1)
+            self._idf_vec[word] = idf
+        if len(self.vocabulary) < self._max_features:
+            self._max_features = len(self.vocabulary)
+        # Sort based on term frequency to match sklearn TfidfVectorizer
+        kbestfeature = dict(sorted(self._idf_vec.items(),
+                                   key=(lambda item:
+                                   self.word_count[item[0]]),
+                                   reverse=True)[0:self._max_features])
+        self.kbestfeature_value = np.array(list(dict(
+            sorted(kbestfeature.items())).values()))
+        kbestfeature_key = np.array(list(dict(
+            sorted(kbestfeature.items())).keys()))
+        self.vocabulary.clear()
+        for i in range(self._max_features):
+            self.vocabulary[kbestfeature_key[i]] = i
+        self._shape = np.shape(self.kbestfeature_value)
 
     def transform(self, x, fit=False):
-        sentence = nest.flatten(x)[0].numpy().decode('utf-8')
-        data = self.vectorizer.transform([sentence]).toarray()
-        if self.selector:
-            data = self.selector.transform(data).astype('float32')
-        return data[0]
+        # Calculate tf at doc level
+        tf = np.zeros(self._max_features, dtype=int)
+        x = nest.flatten(x)[0].numpy().decode('utf-8')
+        token_pattern = re.compile(r"(?u)\b\w\w+\b")
+        tokens = self._word_ngram(token_pattern.findall(x.lower()), self.ngram_range,
+                                  self.stop_words)
+
+        for feature in tokens:
+            if feature in self.vocabulary:
+                feature_idx = self.vocabulary[feature]
+                tf[feature_idx] += 1
+        result = tf * self.kbestfeature_value
+        result = normalize([result], norm=self.norm, copy=False)[0]
+        return result
 
     def output_types(self):
-        return (tf.float32,)
+        return (tf.float64,)
 
     @property
     def output_shape(self):
-        return self.shape
+        return self._shape
 
     def get_weights(self):
-        return {'vectorizer': self.vectorizer,
-                'selector': self.selector,
+        return {'selector': self.selector,
```

Comment:
In `get_weights`, `set_weights`, `get_config`, `set_config`, all fields should be covered. Some self.** are not covered here. The ones that set by the user should be in `get_config` and `set_config`, e.g., norm, stop_words, ngram_range, max_features. The ones inferred from the data should be in `get_weights` and `set_weights`, e.g. shape, vocabulary.